python-spec/src/somacore/query/query.py

```diff
@@ -110,3 +185,378 @@ def __del__(self) -> None:
+
+        # If we can cache join IDs, prepare to add them to the cache.
+        joinids_cached = self._joinids._is_cached(axis)
+        query_columns = column_names
+        if (
+            not joinids_cached
+            and column_names is not None
+            and "soma_joinid" not in column_names
+        ):
+            # If we want to fill the join ID cache, ensure that we query the
+            # soma_joinid column so that it is included in the results.
+            # We'll filter it out later.
+            query_columns = ["soma_joinid"] + list(column_names)
+
+        # Do the actual query.
+        arrow_table = axis_df.read(
+            ids=axis_query.coords,
+            value_filter=axis_query.value_filter,
+            column_names=query_columns,
+        ).concat()
+
+        # Update the cache if needed. We can do this because no matter what
+        # other columns are queried for, the contents of the `soma_joinid`
+        # column will be the same and can be safely stored.
+        if not joinids_cached:
+            setattr(
+                self._joinids,
+                axis.value,
+                arrow_table.column("soma_joinid").combine_chunks(),
+            )
+
+        # Ensure that we return the exact columns the caller was expecting,
+        # even if we added our own above.
+        if column_names is not None:
+            arrow_table = arrow_table.select(column_names)
+        return arrow_table
+
+    def _axisp_inner(
+        self,
+        axis: "_Axis",
+        layer: str,
+    ) -> data.SparseRead:
+        key = axis.value + "p"
+
+        if key not in self._ms:
+            raise ValueError(f"Measurement does not contain {key} data")
+
+        axisp = self._ms.obsp if axis is _Axis.OBS else self._ms.varp
+        if not (layer and layer in axisp):
+            raise ValueError(f"Must specify '{key}' layer")
+        if not isinstance(axisp[layer], data.SparseNDArray):
+            raise TypeError(f"Unexpected SOMA type stored in '{key}' layer")
+
+        joinids = getattr(self._joinids, axis.value)
+        return axisp[layer].read((joinids, joinids))
+
+    @property
+    def _obs_df(self) -> data.DataFrame:
+        return self.experiment.obs
+
+    @property
+    def _ms(self) -> composed.Measurement:
+        return self.experiment.ms[self.measurement_name]
+
+    @property
+    def _var_df(self) -> data.DataFrame:
+        return self._ms.var
+
+    @property
+    def _threadpool(self) -> futures.ThreadPoolExecutor:
+        """Creates a thread pool just in time."""
+        if self._threadpool_ is None:
+            # TODO: the user should be able to set their own threadpool, a la asyncio's
+            # loop.set_default_executor().  This is important for managing the level of
+            # concurrency, etc.
+            self._threadpool_ = futures.ThreadPoolExecutor()
+        return self._threadpool_
+
+
+@attrs.define(frozen=True)
+class AxisQueryResult:
```

Comment:
This is (I believe) a private class and should probably have a underscore-prefixed name

Reply:
ibid.